chia/full_node/block_store.py

```diff
@@ -67,15 +67,16 @@ async def create(cls, db_wrapper: DBWrapper):
         return self
 
     async def add_full_block(self, block: FullBlock, block_record: BlockRecord) -> None:
-        cached = self.block_cache.get(block.header_hash)
+        block_header_hash = block.header_hash
+        cached = self.block_cache.get(block_header_hash)
         if cached is not None:
             # Since write to db can fail, we remove from cache here to avoid potential inconsistency
             # Adding to cache only from reading
-            self.block_cache.remove(block.header_hash)
+            self.block_cache.remove(block_header_hash)
         cursor_1 = await self.db.execute(
             "INSERT OR REPLACE INTO full_blocks VALUES(?, ?, ?, ?, ?)",
             (
-                block.header_hash.hex(),
+                block_header_hash.hex(),
```

Comment:
It's tricky that this property was calling the inner `self.foliage.get_hash()`

What do you think about doing the memoization there?

This issue could be recreated in the future.

Reply:
I'm hesitant to do that, primarily for 2 reasons:
1. I really prefer to solve performance issues by simplifying the code and doing less
2. We definitely pass full blocks across process boundaries, so the caching of the hashes would be inconsistent. I don't think the block store specifically is affected by this though. I think this is all in the main process, but it adds complexity since you still can't rely on this property being cheap.

I think a better long-term solution would be not not have `header_hash` be a property, but be a function called `compute_header_hash()`, to make it obvious to everyone that we're actually computing it.